File: routers/stats.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from sqlalchemy import func

from database import get_db
import models
import schemas
from utils.auth import get_current_admin

router = APIRouter(prefix="/api/stats", tags=["Dashboard Stats"])

@router.get("", response_model=schemas.ApiResponse[schemas.StatsOverview])
def get_dashboard_stats(
    current_user: models.User = Depends(get_current_admin),
    db: Session = Depends(get_db)
):
    """Get overview statistics for admin dashboard."""
    total_pengumuman = db.query(models.Pengumuman).count()
    total_views = db.query(func.coalesce(func.sum(models.Pengumuman.views), 0)).scalar()
    total_pengurus = db.query(models.Pengurus).count()
    total_divisi = db.query(models.Divisi).count()

    latest = db.query(
        models.Pengumuman.id,
        models.Pengumuman.title,
        models.Pengumuman.category,
        models.Pengumuman.date_published,
        models.Pengumuman.views
    ).order_by(models.Pengumuman.created_at.desc()).limit(5).all()

    latest_list = [
        {
            "id": row.id,
            "title": row.title,
            "category": row.category,
            "date_published": row.date_published,
            "views": row.views
        } for row in latest
    ]

    return schemas.ApiResponse(
        success=True,
        data=schemas.StatsOverview(
            totalPengumuman=total_pengumuman,
            totalViews=int(total_views),
            totalPengurus=total_pengurus,
            totalDivisi=total_divisi,
            latestPengumuman=latest_list
        )
    )

File: src/routes/stats.ts
import { Hono } from "hono";
import { Env } from "../types";
import { successResponse } from "../utils/response";
import { authMiddleware } from "../middleware/auth";

const statsRouter = new Hono<{ Bindings: Env }>();

/**
 * GET /api/stats (Protected - Admin)
 */
statsRouter.get("/", authMiddleware, async (c) => {
    const totalPengumuman = await c.env.DB.prepare(
        "SELECT COUNT(*) as count FROM pengumuman"
    ).first<{ count: number }>();

    const totalViews = await c.env.DB.prepare(
        "SELECT SUM(views) as count FROM pengumuman"
    ).first<{ count: number }>();

    const totalPengurus = await c.env.DB.prepare(
        "SELECT COUNT(*) as count FROM pengurus"
    ).first<{ count: number }>();

    const totalDivisi = await c.env.DB.prepare(
        "SELECT COUNT(*) as count FROM divisi"
    ).first<{ count: number }>();

    const latestPengumuman = await c.env.DB.prepare(
        "SELECT id, title, category, date_published, views FROM pengumuman ORDER BY created_at DESC LIMIT 5"
    ).all();

    return c.json(successResponse({
        totalPengumuman: totalPengumuman?.count || 0,
        totalViews: totalViews?.count || 0,
        totalPengurus: totalPengurus?.count || 0,
        totalDivisi: totalDivisi?.count || 0,
        latestPengumuman: latestPengumuman?.results || []
    }));
});

export default statsRouter;
